### ISSUE TITLE
inconsistent context use for import vs export

### ISSUE THREAD

**Issue by blafountain** [schematics/schematics#274]
This might not be a bug, but i guess i'm looking for clarification here related to 'contex' kwargs while import vs export.

I was trying to do a non trivial thing with importing. I found a few references to contexts in the code and in your docs. I was trying to see how that would work to do what I was trying to do. I realized that this only really works for exporting. With the import process there seems like there are a few kwargs to  `convert` but it seems to terminate at `import_loop`, which is confusing to me as the pattern of export is what i would be looking for.

https://github.com/schematics/schematics/blob/development/schematics/transforms.py#L382

as the comment here says, import_loop, context is about default data not a context to pass the whole import chain.
https://github.com/schematics/schematics/blob/development/schematics/transforms.py#L63

After reading thru all of your code and found his PR, which i think is trying to solve the same thing i'm talking about.
https://github.com/schematics/schematics/pull/266

Finally I was able to hack a bit to get around this.
```python
from schematics.models import Model
from schematics.transforms import import_loop
from schematics.types.compound import ModelType, PolyModelType, MultiType

def context_convert(cls, instance_or_dict, context=None, partial=True, strict=False,
            mapping=None):
    def field_converter(field, value, mapping=None):
        try:
            return field.to_native(value, context=context, mapping=mapping)
        except Exception:
            return field.to_native(value, context=context)
    data = import_loop(cls, instance_or_dict, field_converter,
                       partial=partial, strict=strict, mapping=mapping)
    return data

class ContextModel(Model):
  def __init__(self, raw_data=None, context=None, deserialize_mapping=None, strict=True):
      if raw_data is None:
          raw_data = {}
      self._initial = raw_data
      self._data = self.convert(raw_data, context=context, strict=strict, mapping=deserialize_mapping)  

  def convert(self, raw_data, **kw):  
      """
      Converts the raw data into richer Python constructs according to the
      fields on the model
      :param raw_data:
          The data to be converted
      """
      return context_convert(self.__class__, raw_data, **kw)   
```

**Comment by lkraider** [comment]
The context in import_loop is a byproduct of the feature to avoid re-validating data, as you noticed. The naming should be changed, maybe `valid_data`.

**Comment by bintoro** [comment]
Addressed in #359.

Arbitrary data can now be supplied in the `app_data` parameter for both import and export operations. The type of `app_data` can basically be anything (a dict, custom container, ...). The object is accessible at `context.app_data` throughout the (possibly recursive) conversion process.

`context` now represents the entire persistent state of the conversion loop, containing configuration options (such as `strict`) as well as `app_data`.
